src/orchestrator.js

```diff
@@ -80,44 +88,130 @@ function execPreCommands(config) {
+}
+
+function genearateSpecsCommandsForMachines(config, browser) {
+  let specsCommandsOverMachines = [];
+
+  let specs = getListOfSpecs(config, browser);
+  let listOfSpecsOverMachines = splitSpecsOverMachines(specs, config);
+
+  listOfSpecsOverMachines.forEach((listOfspecsPerMachine) => {
     let result = "";
-    end = start + index;
-    specs.slice(start, end).forEach((spec) => {
-      spec = config.specsDockerPath + spec;
-      result = result + "," + spec;
+    listOfspecsPerMachine.forEach((spec) => {
+      result = `${result},${config.specsDockerPath}${spec}`;
     });
-    specsForMachines.push(result.slice(1));
-    start = end;
+    specsCommandsOverMachines.push(result.slice(1));
+  });
+
+  return specsCommandsOverMachines;
+}
+
+function generateSpecsCommandsOverMachinesOrederedByBrowsers(config) {
+  let specsCommandsOverMachinesOrederedByBrowsers = {}; // {'chrome': [ [] , [] , []], 'firefox': [[], [], []]}
+
+  config.browsers.forEach((browser) => {
+    specsCommandsOverMachinesOrederedByBrowsers[browser] =
+      genearateSpecsCommandsForMachines(config, browser);
+  });
+
+  return specsCommandsOverMachinesOrederedByBrowsers;
+}
+
+function _constructCypressCommands(config) {
+  let bashCommands = [];
+  let specsCommandsOverMachinesOrederedByBrowsers =
+    generateSpecsCommandsOverMachinesOrederedByBrowsers(config);
+  let _noOfMachines =
+    specsCommandsOverMachinesOrederedByBrowsers[config.browsers[0]].length;
+
+  for (let i = 0; i < _noOfMachines; i++) {
+    let bashCommand = "exit_code=0";
+
+    let _browsers = i % 2 ? config.browsers : config.browsers.reverse();
+    _browsers.forEach((browser) => {
+      bashCommand = `${bashCommand}; npx cypress run -b ${browser} --headless --spec ${specsCommandsOverMachinesOrederedByBrowsers[browser][i]} || exit_code=$? ; pkill -9 cypress`;
+    });
+
+    bashCommand = `${bashCommand} ; exit $exit_code`;
+    bashCommands.push(bashCommand);
   }
-  return specsForMachines;
+  return bashCommands;
 }
 
 function upConrainters(config) {
   let promises = [];
+  let commands = [];
   let [container_name, command] = ["", ""];
-  let specsForMachines = genearateSpecsForMachines(config);
-  config.browsers.forEach((browser) => {
-    specsForMachines.forEach((specPerMachine) => {
-      container_name = `container_${Math.floor(
-        Math.random() * 100000
-      )}_${browser}_${specsForMachines.indexOf(specPerMachine)}`;
-      command = `timeout --preserve-status ${config.timeout} docker-compose -f ${config.dockerComposePath} run --name ${container_name} ${config.cypressContainerName} npx cypress run -b ${browser} --headless --spec ${specPerMachine}`;
-      console.log(command);
-      promises.push(execa(command));
-    });
+  let bashCommands = _constructCypressCommands(config);
+
+  bashCommands.forEach((cmd) => {
+    container_name = `container_${Math.floor(
+      Math.random() * 100000
+    )}__${bashCommands.indexOf(cmd)}`;
```

Comment:
Just a remark, this `Math.random() * 100000` with `bashCommands.indexOf(cmd)` doesn't guarantee uniqueness, but should be fine.

Reply:
make sense